core/dataset.cpp

```diff
@@ -751,4 +806,38 @@ std::unordered_map<Dim, scipp::index> Dataset::dimensions() const {
   return all;
 }
 
+std::map<typename MasksConstProxy::key_type,
+         typename MasksConstProxy::mapped_type>
+union_or(const MasksConstProxy &currentMasks,
+         const MasksConstProxy &otherMasks) {
+  std::map<typename MasksConstProxy::key_type,
+           typename MasksConstProxy::mapped_type>
+      out;
+
+  for (const auto &[key, item] : currentMasks) {
+    out.emplace(key, item);
+  }
```

Comment:
Would this work here:
```c++
std::map<typename MasksConstProxy::key_type,
￼           typename MasksConstProxy::mapped_type>
out(currentMasks.begin(), currentMasks.end())
```
?

Reply:
Good suggestion, I am also wondering if it's possible to extract the second for-loop into a common function?

Reply:
Possible, but quite ugly: 
```c++
template<class Cm, class Om, class Out>
void common_function(Cm currentMasks, Om otherMasks, Out out) {
    for (const auto &[key, item] : otherMasks) {
      const auto it = currentMasks.find(key);
      if (it != currentMasks.end()) {
        if constexpr (std::is_same_v<Out, Cm>)
          it->second |= item;
        else
          out[key] = it->second | item;
      } else {
        if constexpr (std::is_same_v<Out, Cm>)
          out.set(key, item);
        else
          out.emplace(key, item);
      }
    }
  }
  
std::map<typename MasksConstProxy::key_type,
         typename MasksConstProxy::mapped_type>
union_or(const MasksConstProxy &currentMasks,
         const MasksConstProxy &otherMasks) {
  std::map<typename MasksConstProxy::key_type,
           typename MasksConstProxy::mapped_type>
      out;

  for (const auto &[key, item] : currentMasks) {
    out.emplace(key, item);
  }

  common_function(currentMasks, otherMasks, out);
  return out;
}



void union_or(MasksProxy &&currentMasks, const MasksConstProxy &otherMasks) {
  common_function(currentMasks, otherMasks, currentMasks);
}
```